examples/thing/user-token/main.go

```diff
@@ -0,0 +1,186 @@
+	realm     = flag.String("realm", "", "AM Realm")
+	audience  = flag.String("audience", "/example", "JWT audience")
+	authTree  = flag.String("tree", "iot-tree", "Authentication tree")
+	thingName = flag.String("name", "dynamic-thing", "Thing name")
+	key       = flag.String("key", "", "The Thing's key in PEM format")
+	keyFile   = flag.String("keyfile", "./examples/resources/eckey1.key.pem", "The file containing the Thing's key")
+	cert      = flag.String("cert", "", "The Thing's certificate in PEM format")
+	certFile  = flag.String("certfile", "./examples/resources/dynamic-thing.cert.pem",
+		"The file containing the Thing's certificate if it hasn't been handed directly to the function")
+)
+
+func loadKey() (crypto.Signer, error) {
+	var err error
+	var keyBytes []byte
+	if *key != "" {
+		keyBytes = []byte(*key)
+	} else {
+		keyBytes, err = ioutil.ReadFile(*keyFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	block, _ := pem.Decode(keyBytes)
+	if block == nil {
+		return nil, fmt.Errorf("unable to decode key")
+	}
+	privateKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	return privateKey.(crypto.Signer), nil
+}
+
+func loadCertificates() ([]*x509.Certificate, error) {
+	var err error
+	var certBytes []byte
+	if *cert != "" {
+		certBytes = []byte(*cert)
+	} else {
+		certBytes, err = ioutil.ReadFile(*certFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	block, _ := pem.Decode(certBytes)
+	return x509.ParseCertificates(block.Bytes)
+}
+
+// userTokenThing initialises a Thing with AM and retrieves an access token using OAuth 2.0 device flow.
+// The Thing will register and authenticate with AM and then request a user code.
+// Once the Thing is in procession of a user code, it will direct the user to authorise the token.
+// If successful, the Thing will receive a user token that contains an access token.
+//
+// To create your own certificate, refer to the cert-registration example.
+func userTokenThing() (err error) {
+
+	u, err := url.Parse(*urlString)
+	if err != nil {
+		return err
+	}
+
+	key, err := loadKey()
+	if err != nil {
+		return err
+	}
+
+	// use key thumbprint as key id
+	keyID, err := thing.JWKThumbprint(key)
+	if err != nil {
+		return err
+	}
+
+	certs, err := loadCertificates()
+	if err != nil {
+		return err
+	}
+
+	builder := builder.Thing().
+		ConnectTo(u).
+		InRealm(*realm).
+		WithTree(*authTree).
+		AuthenticateThing(*thingName, *audience, keyID, key, nil).
+		RegisterThing(certs, nil)
+
+	fmt.Printf("Creating Thing %s... ", *thingName)
+	device, err := builder.Create()
+	if err != nil {
+		return err
+	}
+	fmt.Printf("Done\n")
+
+	fmt.Printf("Requesting user code...")
+	userCode, err := device.RequestUserCode("publish")
+	if err != nil {
+		return err
+	}
+	fmt.Println("Done")
+
+	fmt.Printf("Requesting user token... To authorise the Thing, go to \n\n\t%s?user_code=%s\n\n",
```

Comment:
I suppose I could simplify this by using the VerificationURIComplete attribute?

Reply:
Yep.